freeipa/src/main/java/com/sequenceiq/freeipa/flow/freeipa/upscale/action/FreeIpaUpscaleActions.java

```diff
@@ -0,0 +1,482 @@
+
+    @Inject
+    private MetadataSetupService metadataSetupService;
+
+    @Inject
+    private TlsSetupService tlsSetupService;
+
+    @Inject
+    private StackUpdater stackUpdater;
+
+    @Bean(name = "UPSCALE_STARTING_STATE")
+    public Action<?, ?> startingAction() {
+        return new AbstractUpscaleAction<>(UpscaleEvent.class) {
+            @Override
+            protected void doExecute(StackContext context, UpscaleEvent payload, Map<Object, Object> variables) {
+                Stack stack = context.getStack();
+                String operationId = payload.getOperationId();
+                setOperationId(variables, operationId);
+                setInstanceCountByGroup(variables, payload.getInstanceCountByGroup());
+                setRepair(variables, payload.isRepair());
+                LOGGER.info("Starting upscale of stack {}", stack.getId());
+                stackUpdater.updateStackStatus(stack.getId(), DetailedStackStatus.UPSCALE_IN_PROGRESS, "Starting upscale");
+                sendEvent(context, UPSCALE_STARTING_FINISHED_EVENT.selector(), new StackEvent(stack.getId()));
+            }
+        };
+    }
+
+    @Bean(name = "UPSCALE_ADD_INSTANCES_STATE")
+    public Action<?, ?> addInstancesAction() {
+        return new AbstractUpscaleAction<>(StackEvent.class) {
+            @Override
+            protected void doExecute(StackContext context, StackEvent payload, Map<Object, Object> variables) {
+                Stack stack = context.getStack();
+                stackUpdater.updateStackStatus(stack.getId(), DetailedStackStatus.UPSCALE_IN_PROGRESS, "Adding instances");
+
+                List<CloudInstance> newInstances = buildNewInstances(context.getStack(), getInstanceCountByGroup(variables));
+                if (newInstances.isEmpty()) {
+                    List<CloudResourceStatus> list = resourceService.getAllAsCloudResourceStatus(payload.getResourceId());
+                    UpscaleStackRequest<UpscaleStackResult> request = new UpscaleStackRequest<>(
+                            context.getCloudContext(), context.getCloudCredential(), context.getCloudStack(), ResourceLists.transform(list));
+                    UpscaleStackResult result = new UpscaleStackResult(request.getResourceId(), ResourceStatus.CREATED, list);
+                    sendEvent(context, result.selector(), result);
+                } else {
+                    Stack updatedStack = instanceMetaDataService.saveInstanceAndGetUpdatedStack(context.getStack(), newInstances);
+                    List<CloudResource> cloudResources = resourceService.findAllByStackId(stack.getId()).stream()
+                            .map(resource -> resourceConverter.convert(resource))
+                            .collect(Collectors.toList());
+                    CloudStack updatedCloudStack = cloudStackConverter.convert(updatedStack);
+                    UpscaleStackRequest request = new UpscaleStackRequest<UpscaleStackResult>(
```

Comment:
should be: `UpscaleStackRequest<UpscaleStackResult> request = new UpscaleStackRequest<>(`